runtime/lua/vim/treesitter/query.lua

```diff
@@ -254,30 +380,52 @@ function Query:match_preds(match, pattern, source)
     -- Also, tree-sitter strips the leading # from predicates for us.
     local pred_name
     local is_not
-    if string.sub(pred[1], 1, 4) == "not-" then
-      pred_name = string.sub(pred[1], 5)
-      is_not = true
-    else
-      pred_name = pred[1]
-      is_not = false
-    end
 
-    local handler = predicate_handlers[pred_name]
+    -- Skip over directives... they will get processed after all the predicates.
+    if not is_directive(pred[1]) then
+      if string.sub(pred[1], 1, 4) == "not-" then
+        pred_name = string.sub(pred[1], 5)
+        is_not = true
+      else
+        pred_name = pred[1]
+        is_not = false
+      end
 
-    if not handler then
-      a.nvim_err_writeln(string.format("No handler for %s", pred[1]))
-      return false
-    end
+      local handler = predicate_handlers[pred_name]
 
-    local pred_matches = handler(match, pattern, source, pred)
+      if not handler then
+        a.nvim_err_writeln(string.format("No handler for %s", pred[1]))
+        return false
+      end
 
-    if not xor(is_not, pred_matches) then
-      return false
+      local pred_matches = handler(match, pattern, source, pred)
+
+      if not xor(is_not, pred_matches) then
+        return false
+      end
     end
   end
   return true
 end
 
+--- Applies directives against a match and pattern.
+function Query:apply_directives(match, pattern, source)
+  local preds = self.info.patterns[pattern]
+
+  for _, pred in pairs(preds or {}) do
+    if is_directive(pred[1]) then
+      local handler = directive_handlers[pred[1]]
+
+      if not handler then
+        a.nvim_err_writeln(string.format("No handler for %s", pred[1]))
```

Comment:
don't use `nvim_err_writeln`, throw an `error()` instead.

Reply:
I think this was copied from the predicates implementation. Should I change it for both?

Reply:
done!

Reply:
you changed the other case but not this one :]